src/toast/tod/crosstalk.py

```diff
@@ -0,0 +1,473 @@
+# py37+
+# from __future__ import annotations
+
+"""Simulate crosstalk between detectors in ToD.
+
+A typical use of this is that in your pipeline script,
+
+1. Add `add_crosstalk_args(parser)` to your argparse parser to add cli args,
+2. Read & create `op_crosstalk` by adding
+
+
+    .. highlight:: python
+    .. code-block:: python
+
+        if args.crosstalk_matrix is not None:
+            op_crosstalk = OpCrosstalk.read(args)
+
+3. Apply the crosstalk matrix to your data after the ToD is prepared (e.g. signal & noise) by adding
+
+
+    .. highlight:: python
+    .. code-block:: python
+
+        if args.crosstalk_matrix is not None:
+            if comm.comm_world is not None:
+                comm.comm_world.barrier()
+            op_crosstalk.exec(data, "tot_signal")
+            if comm.comm_world is not None:
+                comm.comm_world.barrier()
+
+Note that you need to create your crosstalk matri(x|ces)
+in HDF5 container(s) beforehand. A simple class `SimpleCrosstalkMatrix`
+can assist you to write such file,
+with the simple conventions that
+it has the following datasets:
+
+1. names: ASCII names of your detectors, say of length ``n``
+2. data: ``n`` by ``n`` float64 array of your crosstalk matrix ``m``, where you expect
+
+    ToD_crosstalked = m @ ToD_original
+
+    (Assuming row-major as is standard in Python.)
+
+Note that each crosstalk matrix is dense,
+and when multiple matrices are supplied, they are effectively block-diagonal
+(where each given dense matrix is a block.)
+I.e. in order to provide a block-diagonal crosstalk matrix,
+each block (together with the names of the detectors in that block)
+should write to a seperate HDF5 file to avoid unnessary computation.
+"""
+
+import argparse
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import h5py
+import numpy as np
+
+try:
+    from numba import jit
+except ImportError:
+    jit = None
+
+from ..mpi import get_world
+from ..op import Operator
+from ..utils import Logger
+
+if TYPE_CHECKING:
+    from typing import List  # Optional
+
+    import toast
+
+# py37+
+# COMM: Optional[toast.mpi.Comm]
+# PROCS: int
+# RANK: int
+COMM, PROCS, RANK = get_world()
```

Comment:
Resolved 1st one in 5c1c5f71.

Actually I think currently it can handle different crosstalk matrices for different observations. The logic is mainly encoded in ln 440-453 for MPI case and ln 376-388 for serial case:

```py
        for idx_crosstalk_matrix in range(self.n_crosstalk_matrices):
            crosstalk_matrix = self._get_crosstalk_matrix(idx_crosstalk_matrix)
            ...
            for obs in data.obs:
                ...
                # all ranks need to check this as they need to perform the same action
                detectors_set = set(tod.detectors)
                if not (names_set & detectors_set):
                    logger.info(
                        f"Crosstalk: skipping tod {tod} as "
                        "it does not include detectors from crosstalk matrix "
                        f"with these detectors: {names}."
                    )
                    continue
                elif not (names_set <= detectors_set):
                    raise ValueError(
                        f"Crosstalk: tod {tod} only include some detectors "
                        f"from the crosstalk matrix with these detectors: {names}."
                    )
```

So (if I think correctly...),

1. all crosstalk matrices needed in a pipeline script should be provided by `--crosstalk-matrix path1 path2 ...`, including those from any telescopes, etc.
2. the `read` method will distribute them evenly in the whole world, just to avoid the situation that there might be a lot and overwhelm the memory in the node of "master" process.
3. per observation, by the logic quoted above, it loops over all crosstalk matrices first, then all obs, (so each crosstalk matrix is only MPI-communicated once)
4. then it compares between the detectors of the crosstalk matrix and the detectors of the ToD of current obs. If they don't overlap, skip it.

I think the main reason for this is that the relationship of crosstalk matrix and obs are not clear. i.e. if we need to restrict the crosstalk matrix to the group that needs it, that we need to provide a way to convey that.

But if we can do that then the "crosstalk matrix" loop is shorter, avoiding it to do unnecessary check if we know ahead of time that some matrices has nothing to do with these observations.

So **to summarize**, I think the current implementation can still handles the most general case, but it can be improved to avoid unnecessary runtime checks if the information can be provided ahead of time how some matrices to be mapped to some telescope / groups.

Also, may be it's worth mention here that ultimately we want to do cross-wafer crosstalk (however there's no cross-telescope crosstalk.) So may be we can cover that case (plus the improvement mentioned here) when porting to TOAST 3?

Reply:
@tskisner, it is explained here when the world communicator is used. In short, the 2 classes `OpCrosstalk` and `SimpleCrosstalkMatrix` handles MPI very differently. `OpCrosstalk` as a toast operator uses the MPI communicator from the data (`comm = tod.grid_comm_col`). But `SimpleCrosstalkMatrix` is not a toast operator (but is part of the attributes of OpCrosstalk: `OpCrosstalk.crosstalk_matrices: List[SimpleCrosstalkMatrix]`), and the matrices is loaded the moment `op_crosstalk` object is created, which does not yet know what `data` it is going to `exec` on.

In the current design, eventually all rank are going to see all crosstalk matrices (`crosstalk_matrix = self._get_crosstalk_matrix(idx_crosstalk_matrix)`), the only reason to need MPI here (besides perhaps an initial broadcast) is such that the multiple crosstalk matrices are not all in the root process, which also has an advantage of multiple ranks doing the parallel read initially.

Reply:
Resolved after an internal discussion on our Tuesday meeting. The notes here and the additional paragraph in https://github.com/ickc/toast/blob/bbcf03b28e350895bc42e706981e79b86b163073/src/toast/tod/crosstalk.py#L213-L218 explains the decision of using the world communicator there.